lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -36,27 +34,45 @@ def execute(self):
             # We were not given any remotes on the command line, build the list
             # from the tools group directory.
             remotes = self.remote(self.tool_group_dir)
-
+            if not remotes:
+                # FIXME:  This is a weird case -- the group directory exists,
+                #  but contains no remotes. This is the result of clearing the
+                #  tools from a group which has already been cleared, because
+                #  the first clear does not remove the group directory.  This
+                #  behavior might be desirable for the `default` directory, but
+                #  it seems wrong for custom groups.
+                self.logger.error(f'No such group "{self.context.group}".')
+                return 1
+
+        tool_not_found = bool(self.context.name)  # Can't not find if not specified
         for remote in remotes:
             tg_dir_r = self.tool_group_dir / remote
             if not tg_dir_r.exists():
                 self.logger.warn(
-                    'The given remote host, "%s", is not a directory in' " %s.",
-                    remote,
-                    self.tool_group_dir,
+                    'No remote host "%s" in group %s.', remote, self.context.group,
                 )
                 continue
 
             if self.context.name:
-                names = self.context.name
+                names = [self.context.name]
             else:
                 # Discover all the tools registered for this remote
                 names = self.tools(tg_dir_r)
+                if not names:
+                    # FIXME:  this is another odd case -- the remote subdirectory
+                    #  exists, but it's empty.  (We'll remove it below.)
+                    self.logger.warn(
+                        'No tools in group "%s" on host "%s".',
+                        self.context.group,
+                        remote,
+                    )
```

Comment:
Is this a case the *user* really even needs to know about? As a "real" `logger.warn` I might be in favor, but as these commands (ridiculously) use a file-streamed logger as user feedback, this seems excessive. It's almost certainly an internal error, it's not harmful, and we're about to clean it up.

Reply:
Unfortunately, I'm building an edifice on a rickety foundation, here.

It should be that this case never arises, and, therefore, there should be no need to report it to anybody.  Yet, the code (silently, prior to my change) defends itself against the situation.  And, in the case where there is only one host (or where all the hosts are in the same state), the command will exit silently, which is something that I'm trying to ensure never happens:  I want it to either report success or report errors -- or both -- but it should never report nothing.

So, assuming that the user remembers (or at least _did_) register tools for that host, I think it's worth having the tool note that there are no tools for that host, now (i.e., that the "clear" operation was aware of the host, and tried to clear the requested tools from it, but determined that they weren't there).

[But...I was hoping that we could discuss this, which is why I put the `FIXME` there....  🙂]

Reply:
That basic UNIX command philosophy is targeted at pipeline components, not at human users.

Given that the command in question doesn't take input from `stdin` nor produce output which is designed to be passed on to a subsequent pipeline stage, I'd say this command falls into the "human interface" bucket, where "silence is ambiguous".

I would be willing to put this output under a `-v` flag if we had one, but, given that we don't, I would prefer to just put it in the user's view in the rare case where it's actually appropriate.

What I'm more interested in pondering here is whether we should be contemplating this case at all.  E.g., should we just be skipping this `remote` completely like we do in the previous `if`-test, if the directory _doesn't_ exist?  The fact that this test triggered indicates that the subdirectory is in a "bad" state...should we clean it up or ignore it?  (Perhaps we should tell the user to clean it up?)